src/messaging.ts

```diff
@@ -0,0 +1,144 @@
+import { fromEvent, Observable } from 'rxjs'
+import { filter, map, take } from 'rxjs/operators'
+import { UpdateExtensionSettingsArgs } from './context'
+
+/**
+ * One end of the client-page connection. The browser extension is a client, and
+ * the web app is a page.
+ */
+export type Source = 'Client' | 'Page'
+
+/** Useful for building tagged union types. */
+interface Tagged<T> {
+    type: T
+}
+
+/**
+ * Enumeration of all the types of messages that could be sent between the page
+ * and client.
+ */
+type MessageType = 'Ping' | 'EditSettings' | 'GetSettings' | 'Settings'
+
+/**
+ * The first message sent between the page and client, and used to detect
+ * presence of each other.
+ */
+export interface Ping extends Tagged<MessageType> {
+    type: 'Ping'
+}
+
+/**
+ * Intent to change a settings value. This is sent from the page to the client.
+ */
+export interface EditSettings extends Tagged<MessageType> {
+    type: 'EditSettings'
+    edit: UpdateExtensionSettingsArgs
+}
+
+/**
+ * A request for the client to send the latest settings back. This is sent from
+ * the page to the client.
+ */
+export interface GetSettings extends Tagged<MessageType> {
+    type: 'GetSettings'
+}
+
+/**
+ * A notification that the client settings have been updated. This is sent from
+ * the client to the page.
+ */
+export interface Settings extends Tagged<MessageType> {
+    type: 'Settings'
+    settings: string
+}
+
+/** A message that is passed between the client and page. */
+type Message = Ping | EditSettings | GetSettings | Settings
+
+/** A connection to the client. */
+export interface ClientConnection {
+    /** Listens for the latest client settings. */
+    onSettings: (callback: (settings: string) => void) => void
```

Comment:
Is there a reason this isn't simply an Observable? I would use handler registration for request/response, but this seems more like a notification, so it just looks like a hand-rolled event emitter with fuzzy semantics (for example, does calling it twice override the handler or register two handlers?)

Reply:
I switched to vscode-jsonrpc for request/response, so this interface inherits its semantics (I think calling `onNotification` for the same method twice replaces the old handler with the new one).

Reply:
What was the reasoning behind using vscode-jsonrpc? From my experience working with it I found it a terrible overly-complex pseudo-typesafe inflexible hard-to-debug node-dependent unidiomatic implementation of an RPC library, so I would stay away from it as much as possible, especially if all we do is communicate with a WebWorker

Reply:
Mostly because I'm familiar with it. The version in the sourcegraph package works in the browser. What do you prefer to use?

Reply:
`vscode-jsonrpc` implements a lot of things that we don't need specically the LSP JSON RPC transport messages with headers and bodies. If all we need is event emitters, that's what Observables/Subjects are, which are composable in opposite to callbacks.

For request/response handlers I would probably ask myself first if I need dynamic registration. Especially if you cannot actually register multiple handlers but a new handler replaces the previous, it would make more sense to me to pass a handler object when creating the connection that contains methods for all  functions that can be handled. You could mutate that object if you need to.

If dynamic registration is needed, I made a pretty detailed proposal on what a better vscode-jsonrpc API could look like over a year ago: https://github.com/Microsoft/vscode-languageserver-node/issues/182

Reply:
Quoteworthy response:

> I hope this explains why the implementation and design is like it is today. Would I implement it different today if I would start from scratch: Yes. Do I think that the current implementation and design is so bad that it is worth to do a rewrite and break all the user. No.

Since we are starting from scratch...

Reply:
Gotcha, passing in the handlers would make the code a bit more safe.